torch/csrc/jit/tensorexpr/block_codegen.h

```diff
@@ -0,0 +1,147 @@
+#pragma once
+
+#include <unordered_map>
+#include <unordered_set>
+
+#include <ATen/ATen.h>
+#include <torch/csrc/jit/resource_guard.h>
+#include <torch/csrc/jit/tensorexpr/analysis.h>
+#include <torch/csrc/jit/tensorexpr/codegen.h>
+#include <torch/csrc/jit/tensorexpr/ir.h>
+#include <torch/csrc/jit/tensorexpr/ir_printer.h>
+#include <torch/csrc/jit/tensorexpr/ir_visitor.h>
+#include <torch/csrc/jit/tensorexpr/unique_name_manager.h>
+
+namespace torch {
+namespace jit {
+namespace tensorexpr {
+
+// A class that analyzes the given program relevant for Block backend.
+class BlockAnalysis : public IRVisitor {
+ public:
+  bool is_buf_store_target(const Buf* buf) const {
+    return store_targets_.count(buf) > 0;
+  }
+
+  const std::unordered_set<const Buf*>& loads() const {
+    return loads_;
+  }
+
+  const std::unordered_set<const Buf*>& stores() const {
+    return store_targets_;
+  }
+
+  int block_size() const {
+    return block_size_;
+  }
+
+  bool areBufsInMap(const std::unordered_set<const Buf*>& bufs) const;
+
+  const Buf* getMultiDimBuf(const Buf* buf) const;
+
+  std::string getInputName(const Buf* buf) const;
+
+  std::string getFlatInputName(const Buf* buf) const {
+    return getInputName(buf) + "_flat";
+  }
+
+  std::unordered_map<std::string, const Buf*> getBufferMap() const {
+    return map_input_to_tensor_bufs_;
+  }
+
+ private:
+  void visit(const Store* v) override;
+  void visit(const Load* v) override;
+  void visit(const For* v) override;
+
+  std::unordered_map<std::string, const Buf*> map_input_to_tensor_bufs_;
+  std::unordered_set<const Buf*> store_targets_;
+  std::unordered_set<const Buf*> loads_;
+  int block_size_ = 32;
+};
+
+// A class that overrides the underlying IRPrinter to produce Block.
+class BlockPrinter : public IRPrinter {
+ public:
+  explicit BlockPrinter(std::ostream* os, const BlockAnalysis* block_analysis)
```

Comment:
Nit picking: explicit is not useful here.

Reply:
thanks, took it out